scripts/gen_charm_internal_version.sh

```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+# This file consistently generates a hash, whether there's uncommited code or not
+# and writes it into the `charm_internal_version` file
+
+if git --version > /dev/null; then
+    # Compute base files
+    BASE=$(git ls-files -s)
+    # Compute diff files
+    DIFF=$(git diff --raw)
+    # Compute staged files
+    STAGED=$(git diff --raw --staged)
+
+    HASH=$(echo $BASE $DIFF $STAGED | git hash-object --stdin | cut -c 1-8)
```

Comment:
why is the `cut` necessary here? why not store the whole string?

Reply:
We discussed about it with @MiaAltieri and to avoid too long strings/messages in statuses, we decided to go to a shorter version.
It's not really important in itself, just makes things easier to read and understand
It still keeps something like 4B possibilities so there's a long way before we have a collision (and it's what git does too to generate the short id of a commit)